src/clj/fluree/db/reasoner/core.cljc

```diff
@@ -0,0 +1,151 @@
+(ns fluree.db.reasoner.core
+  (:require [clojure.core.async :as async :refer [alts! go]]
+            [fluree.db.flake :as flake]
+            [fluree.db.util.core :as util]
+            [fluree.db.util.log :as log]
+            [fluree.db.util.async :refer [go-try <?]]
+            [fluree.db.reasoner.resolve :as resolve]
+            [fluree.db.json-ld.transact :as transact]
+            [fluree.db.fuel :as fuel]
+            [fluree.json-ld :as json-ld]
+            [fluree.db.reasoner.graph :refer [task-queue add-rule-dependencies]]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+(defn reasoned-flake?
+  "Returns truthy if the flake has been generated by reasoner"
+  [flake]
+  (-> flake flake/m :reasoned))
+
+(defn non-reasoned-flakes
+  "Takes a sequence of flakes and removes any flakes which are reasoned.
+
+  This is primarily used to remove reasoned flakes from commits."
+  [flakes]
+  (remove reasoned-flake? flakes))
+
+(defn reasoned-flakes
+  "Takes a sequence of flakes and keeps only reasoned flakes"
+  [flakes]
+  (filter reasoned-flake? flakes))
+
+(defn schedule
+  "Returns list of rule @id values in the order they should be run.
+
+  If optional result-summary is provided, rules that don't need to be
+  re-run will be filtered out.
+
+  A result summary is list/set of the rule dependency patterns which
+  match newly created Flakes from the last run. When the result-summary
+  is empty, no rules will be left to run, but based on the dependencies
+  it is possible no rules will be left to run even if the result-summary
+  is non-empty"
+  ([rules]
+   (task-queue rules))
+  ([rules result-summary]
+   (task-queue rules result-summary)))
+
+(defn reasoner-stage
+  [db fuel-tracker txn-id author-did rule-id full-rule]
+  (go-try
+    (let [tx-state      (transact/->tx-state db txn-id author-did rule-id)
+          parsed-txn    (:rule-parsed full-rule)
+          _             (when-not (:where parsed-txn)
+                          (throw (ex-info (str "Unable to execute reasoner rule transaction due to format error: " (:rule full-rule))
+                                          {:status 400 :error :db/invalid-transaction})))
+
+          flakes-ch     (transact/generate-flakes db fuel-tracker parsed-txn tx-state)
+          fuel-error-ch (:error-ch fuel-tracker)
+          chans         (remove nil? [fuel-error-ch flakes-ch])
+          [flakes] (alts! chans :priority true)]
+      (when (util/exception? flakes)
+        (throw flakes))
+      flakes)))
+
+(defn execute-reasoner-rule
+  [db rule-id reasoning-rules fuel-tracker {:keys [txn-id author-did] :as tx-state}]
+  (go-try
```

Comment:
I think there could be significant benefits to this, but I'm assuming you mean executing them in a pipeline... so happy to implement that when we've got everything working.

Reply:
Right. I said "parallel" because once we get things in a pipeline, we can tweak the buffer sizes on the channels involved and get behavior very close to parallel, sometimes. There lots of caveats to that, of course.